### ISSUE TITLE
Add auto-route options

### ISSUE THREAD

**Issue by danrl** [danrl/luci-proto-wireguard#2]
Fixing issue #1 removed the automatic creation of routes.
However, the use case of routes that come up automatically with the interface still exists.
We should be able to add routes to the interface configuration.

**Comment by bol-van** [comment]
The only autoroutes I see are related to configured ip addresses.
Such as 192.168.5.1/24 will see route  "192.168.5.0/24 dev wg src 192.168.5.1". This is normal.
What we still miss is proto_add_host_dependency if endpoint ip is present. It prevents us from cutting ourselves from the server if we want default route.
See example in l2tp script.

**Comment by bol-van** [comment]
About routes.. We can add them as static routes, and this is fully conformant with netifd.
But for convenience we may want to have "Use default gateway", "Use gateway metric" controls - the same as pptp configurator shows.

**Comment by danrl** [comment]
> The only autoroutes I see are related to configured ip addresses.
Well, usually one likes to route every network from AllowedIPs I guess. Otherwise we may end up with split routes. What is the use accepting packets from, let's say, fd00:a:b:c::/48 but not wanting to route the answer packets back through the tunnel? Makes not much sense to me.
 Example:
`AllowedIPs=fd00:a:b:c::/48, fd00:f:f:f::/64`
For example with`fd00:f:f:f::1/64` being the interface address, the interface route
`fd00:f:f:f::/64 src fd00:f:f:f::1/64` will be automatically created. But for `fd00:a:b:c::/48`, a network somewhere behind the other side of the tunnel, we will need to have a route.

As I understand, it is discussed on the Wireguard ML if these will be created automatically in the future. But for now, we have to create them with wireguard.sh

> What we still miss is proto_add_host_dependency if endpoint ip is present. It prevents us from cutting ourselves from the server if we want default route.

That's right. still missing. Will fix it.

> We can add them as static routes, and this is fully conformant with netifd.
Will static routes be added every time the interface comes up? E.g. when users restart the interface often?

**Comment by bol-van** [comment]
> Will static routes be added every time the interface comes up? E.g. when users restart the interface often?
Yes, netifd handles them. I'm on such config and it works flawlessly.

> Well, usually one likes to route every network from AllowedIPs I guess. 
Not always. Here is my usage case.
On server side I set allowed IPs to 192.168.4.0/24 and route to 192.168.4.0/24.  Route=Allowed IP like you say.
But on client side I want to forward very large ip group to wg server. I use ipset, iptables mark , ip rule and separate routing table for wg interface (0.0.0.0/0 dev wg).
Ipset contain ~30000 ip addresses and dynamically changes. Addresses there are IPs blocked by censor and they need to be redirected.
What I write to AllowedIPs ? I cant put there 30K ips and reconfig on any ipset renew. I put there 0.0.0.0/0 because I want to accept any source IP from server.

**Comment by danrl** [comment]
Ok, that is an interesting use case.
Given that static routes work fine, I think we can skip the whole auto-route part in general.

So only thing left is `proto_add_host_dependency` I guess.

**Comment by bol-van** [comment]
I guess it would be wise to offer a choice.
In many cases user will want route=AllowedIP.
How about checkbox ?

**Comment by danrl** [comment]
Current state:

- Checkbox for automatic route creation (seems to work)
- Endpoint dependency is managed now. (I don't know if that really works for hostNAMES as well as for IP addresses. netifd does not complain, but how to verify?

Please test the code and let me know of any show stoppers. Will run some tests, too. Hopefully this can be pushed upstream soon.

**Comment by bol-van** [comment]
> Endpoint dependency is managed now. (I don't know if that really works for hostNAMES as well as for IP addresses. netifd does not complain, but how to verify?

Use default route. If it does not cut itself then it works. Check routing table. Server IP specific route should be there

**Comment by danrl** [comment]
I thought of something more deeper in the code, e.g. a line of code that resolves the name and makes sure both, IPv4 and IPv6 addresses are all taken care of. Sure I can see the result, but that does not tell me much about edge cases. For example multiple A and AAAA RR for a hostname.

**Comment by bol-van** [comment]
This is not your responsibility. If netifd added server route then it received it. Other protocol handlers do the same job

**Comment by bol-van** [comment]
wait.. look in l2tp.sh

```
json_get_var server server                                           
host="${server%:*}"                                                  
for ip in $(resolveip -t 5 "$host"); do                              
 ( proto_add_host_dependency "$interface" "$ip" )             
 serv_addr=1                                                  
done
[ -n "$serv_addr" ] || {                                             
 echo "Could not resolve server address" >&2
 sleep 5
 proto_setup_failed "$interface"
 exit 1
}
```

**Comment by bol-van** [comment]
also see function definition
```
proto_add_host_dependency() {
 local interface="$1"
 local host="$2"
 local ifname="$3"
```
3rd parameter contain logical interface name tunnel depends on
for some interface types its even documented in wiki
there's interface parameter "tunlink". its passed as ifname

**Comment by danrl** [comment]
Should be working now, starting from 03fd198
It's late, so I did just a short test.

**Comment by bol-van** [comment]
I tested with default route. proto_add_host_dependency called with right ip. but route not added, we're cut from the server.  Is it working on your side ?

**Comment by danrl** [comment]
It's working on my side. However, I had to delete the interface, removing all it's configuration, and create a new one. Just overwriting the files with new ones from this repo did not work.

Here is what I did:
- Remove wg interface (delete)
- Make sure all relevant data is gone in `/etc/config/network`
- update files by `git pull` and `./install.sh`
- reboot
- create new wg interface

**Comment by bol-van** [comment]
Do you see server specific route in "ip route" output ?
I rebooted my router and there was no such route.

```
config interface 'wgkk'
        option proto 'wireguard'
        option auto '0'
        option private_key '<<pvk>>'
        option preshared_key '<<psk>>'
        option listen_port '16'
        option mtu '1370'
        list addresses '192.168.254.66/27'
        list addresses 'fdff:ffff:fffe::2/64'
        option metric 4

config wireguard_wgkk
        option public_key '<pubk>'
        list allowed_ips '0.0.0.0/0'
        list allowed_ips '::/0'
        option persistent_keepalive '20'
        option route_allowed_ips 'enabled'
        option endpoint_host 's.s.s.s'
        option endpoint_port '16'

```

**Comment by danrl** [comment]
Here is my config, quite similar to yours.

```
config interface 'foo'
	option proto 'wireguard'
	option private_key 'PRIVKEY'
	option preshared_key 'PSK'
	list addresses 'fd00:13:37::1/64'
config wireguard_foo
	option public_key 'CQlbo7I9It1clntLATRGz5DBbgE5XQImurw7d688/FQ='
	option route_allowed_ips 'enabled'
	option endpoint_host 'm.sealand.io'
	option endpoint_port '500'
	option persistent_keepalive '18'
	list allowed_ips 'fd00:a:b:c::/48'
	list allowed_ips '172.16.17.18/24'
```

Works as far as I can tell.

```
root@OpenWrt:~# ip -6 r s
default from 2001:4ca0:0:f223::/64 via fe80::da67:d9ff:fe6e:a242 dev eth1  proto static  metric 512  pref medium
2001:67c:26f4:64::72 via fe80::da67:d9ff:fe6e:a242 dev eth1  proto static  metric 512  pref medium <--- THERE IT IS!
[...]
```

Legacy IP, which I don't care about much. But nice if it works, too.

```
root@OpenWrt:~# ip -4 r s
default via 10.180.63.254 dev eth1  proto static  src 10.180.51.67 
10.180.48.0/20 dev eth1  proto kernel  scope link  src 10.180.51.67 
10.180.63.254 dev eth1  proto static  scope link  src 10.180.51.67 
192.168.1.0/24 dev br-lan  proto kernel  scope link  src 192.168.1.1 
193.160.39.72 via 10.180.63.254 dev eth1  proto static <--- THERE IT IS!
```


What might have gone wrong: You did not use an IP address "s.s.s.s" but a hostname, AND the resolver was not yet up when netifd tried to bring up the interface. Possible?

**Comment by danrl** [comment]
I added some robustness in f93ffa3beb1290c2f378caa819e8a2e64c9274a6, interface setup will fail if any endpoint of any peer fails to resolve. I think this is OK for usual VPNs, but with wg one might run many peers and endpoints, and the interface not coming up because a single endpoint of one peer does not resolve is not what I like. However, that is probably the only way to make host dependency work reliably. Still a compromise.

Hope that solves your problem?

**Comment by bol-van** [comment]
It adds server route to the wrong interface.  eth instead of l2tp. I tried adding hard-coded 3rd parameter to add_host_dependency (tunlink) but it didnt help.

**Comment by bol-van** [comment]
Sometimes it doesnt add server route at all. I checked that ubus is called with valid parameters. Its not your fault. There's always workaround to manually add route

**Comment by danrl** [comment]
I will close this issue then.
Next couple of days I will run some more tests with different configurations and devices. If all tests pass, I will push the changes to upstream soon.

Thank you for the help and efforts and the intense testing. Very much appreciated!
